File: LeetCodeProblemsCMake/Problems/LeetCode3147TakingMaximumEnergyFromTheMysticDungeon.cpp
//
// Created by saeed on 10/10/25.
//

#include "LeetCode3147TakingMaximumEnergyFromTheMysticDungeon.h"

#include <limits.h>
using namespace std;
int getEnergy(vector<int>& energies, vector<int> &memo,int length,int  k,int index) {
    if (memo[index]!=-1)
        return memo[index];
    if (index<k) {
        memo[index]=energies[index];
        return energies[index];
    }
    memo[index]=max(energies[index], energies[index]+getEnergy(energies,memo,length,k,index-k));
    return memo[index];
}
int LeetCode3147TakingMaximumEnergyFromTheMysticDungeon::maximumEnergy( vector<int> &energy, int k) {
    int length=static_cast<int> (energy.size());
    vector<int> prefixEnergy(length,-1);
    int maxEnergy=INT_MIN;
    for(int i=length-1;i>=length-k;i--) {
        maxEnergy=max(maxEnergy,getEnergy(energy,prefixEnergy,length,k,i));
    }
    return maxEnergy;
}

File: LeetCodeProblemsGo/LeetCode3147TakingMaximumEnergyFromTheMysticDungeon/LeetCode3147TakingMaximumEnergyFromTheMysticDungeon.go
package LeetCode3147TakingMaximumEnergyFromTheMysticDungeon

import "math"

func maximumEnergy(energy []int, k int) int {
	length := len(energy)
	suffixSums := make([]int, length)
	maxEnergy := math.MinInt
	for i := length - 1; i >= 0; i-- {
		suffixSums[i] = energy[i]
		if i+k <= length-1 {
			suffixSums[i] += suffixSums[i+k]
		}
		maxEnergy = max(suffixSums[i], maxEnergy)
	}
	return maxEnergy
}
